File: backend/app/core/demo_data.py
from __future__ import annotations

import math
from datetime import date, timedelta
from typing import Any


DEMO_TICKERS = {
    "AAPL": ("Apple Inc.", "Technology", "NASDAQ", 212.4),
    "MSFT": ("Microsoft Corporation", "Technology", "NASDAQ", 498.3),
    "NVDA": ("NVIDIA Corporation", "Technology", "NASDAQ", 144.9),
    "AMD": ("Advanced Micro Devices, Inc.", "Technology", "NASDAQ", 164.2),
    "SPY": ("SPDR S&P 500 ETF Trust", "Benchmark", "NYSEARCA", 612.1),
    "QQQ": ("Invesco QQQ Trust", "Benchmark", "NASDAQ", 545.6),
}


def demo_signals() -> list[dict[str, Any]]:
    return [demo_summary(symbol)["signal"] for symbol in ["NVDA", "MSFT", "AMD", "AAPL"]]


def demo_summary(symbol: str) -> dict[str, Any]:
    symbol = symbol.upper()
    name, sector, exchange, base_price = DEMO_TICKERS.get(
        symbol, (f"{symbol} Corporation", "Watchlist", "NASDAQ", 100.0)
    )
    history = _price_history(symbol, base_price)
    latest = history[-1]
    probability = {
        "NVDA": 0.63,
        "MSFT": 0.59,
        "AMD": 0.56,
        "AAPL": 0.54,
    }.get(symbol, 0.52)

    events = demo_events(symbol)
    signal = {
        "id": f"demo-{symbol.lower()}-signal",
        "symbol": symbol,
        "signal_date": latest["date"],
        "horizon": "5 trading days",
        "probability_outperform_spy": probability,
        "confidence": "low",
        "score": round((probability - 0.5) * 2, 4),
        "drivers": [
            "Demo momentum is positive versus the watchlist baseline.",
            "Recent educational sample news skews slightly positive.",
        ],
        "risks": [
            "This is demo data. Refresh daily data before using the research workflow.",
            "Model uncertainty is high until real source evidence is cached.",
        ],
        "evidence": [
            {
                "type": event["event_type"],
                "summary": event["summary"],
                "url": event["source_url"],
                "sentiment": event["sentiment"],
                "confidence": event["confidence"],
            }
            for event in events[:2]
        ],
    }

    return {
        "symbol": symbol,
        "name": name,
        "sector": sector,
        "exchange": exchange,
        "latest_price": latest["close"],
        "latest_date": latest["date"],
        "return_1d": 0.006,
        "return_5d": 0.024,
        "return_20d": 0.071 if symbol in {"NVDA", "MSFT"} else 0.034,
        "relative_strength_spy": 0.028 if symbol in {"NVDA", "MSFT"} else -0.004,
        "volume_ratio_20d": 1.32,
        "signal": signal,
        "price_history": history,
        "events": events,
        "what_changed": [
            "Demo price action shows a constructive daily-close setup.",
            "The signal is awaiting real price, filing, and news ingestion.",
        ],
        "why_it_matters": [
            "StockTek ranks tickers by probability, confidence, and evidence quality.",
            "Every real signal should be checked against source links and backtests.",
        ],
        "risks": signal["risks"],
        "next_watch": [
            "Run Refresh Daily Data after market close.",
            "Check whether the signal still beats SPY in backtesting.",
        ],
        "is_demo": True,
    }


def demo_events(symbol: str) -> list[dict[str, Any]]:
    symbol = symbol.upper()
    today = date.today()
    return [
        {
            "id": f"demo-{symbol.lower()}-earnings",
            "symbol": symbol,
            "signal_date": today - timedelta(days=1),
            "event_type": "earnings",
            "sentiment": 0.4,
            "confidence": 0.55,
            "source_url": "https://www.sec.gov/edgar/search/",
            "summary": f"{symbol} demo earnings and filing watch item",
        },
        {
            "id": f"demo-{symbol.lower()}-news",
            "symbol": symbol,
            "signal_date": today - timedelta(days=2),
            "event_type": "general",
            "sentiment": 0.2,
            "confidence": 0.45,
            "source_url": "https://www.gdeltproject.org/data.html",
            "summary": f"{symbol} demo market news item from free source discovery",
        },
    ]


def _price_history(symbol: str, base_price: float) -> list[dict[str, Any]]:
    today = date.today()
    points: list[dict[str, Any]] = []
    seed = sum(ord(char) for char in symbol)
    for index in range(60):
        days_back = 59 - index
        current_date = today - timedelta(days=days_back)
        wave = math.sin((index + seed) / 7.0) * 0.035
        trend = index * 0.0018
        close = round(base_price * (1 + trend + wave), 2)
        points.append({"date": current_date, "close": close, "volume": 10000000 + index * 35000})
    return points

File: frontend/src/lib/demoData.ts
import type {
  BacktestResponse,
  HealthResponse,
  Signal,
  Ticker,
  TickerSummary,
} from "../types/api";

const today = new Date();
const isoDate = (daysAgo = 0) => {
  const value = new Date(today);
  value.setDate(value.getDate() - daysAgo);
  return value.toISOString().slice(0, 10);
};

const tickers: Ticker[] = [
  { symbol: "AAPL", name: "Apple Inc.", sector: "Technology", exchange: "NASDAQ", active: true },
  { symbol: "MSFT", name: "Microsoft Corporation", sector: "Technology", exchange: "NASDAQ", active: true },
  { symbol: "NVDA", name: "NVIDIA Corporation", sector: "Technology", exchange: "NASDAQ", active: true },
  { symbol: "AMD", name: "Advanced Micro Devices, Inc.", sector: "Technology", exchange: "NASDAQ", active: true },
  { symbol: "SPY", name: "SPDR S&P 500 ETF Trust", sector: "Benchmark", exchange: "NYSEARCA", active: true },
];

export const demoHealth: HealthResponse = {
  status: "demo",
  educational_research_only: true,
  latest_price_date: null,
  counts: {
    tickers: tickers.length,
    price_bars: 0,
    source_documents: 0,
    event_signals: 0,
    feature_snapshots: 0,
    signals: 0,
    backtest_runs: 0,
  },
};

export const demoWatchlist = tickers;

export const demoSignals: Signal[] = [
  signal("NVDA", 0.63, "medium"),
  signal("MSFT", 0.59, "medium"),
  signal("AMD", 0.56, "low"),
  signal("AAPL", 0.54, "low"),
];

export function demoSummary(symbol: string): TickerSummary {
  const selected = tickers.find((item) => item.symbol === symbol) ?? tickers[0];
  const matchedSignal = demoSignals.find((item) => item.symbol === selected.symbol) ?? signal(selected.symbol, 0.52, "low");
  return {
    symbol: selected.symbol,
    name: selected.name,
    sector: selected.sector,
    exchange: selected.exchange,
    latest_price: selected.symbol === "NVDA" ? 144.9 : selected.symbol === "MSFT" ? 498.3 : 212.4,
    latest_date: isoDate(),
    return_1d: 0.006,
    return_5d: selected.symbol === "NVDA" ? 0.041 : 0.019,
    return_20d: selected.symbol === "NVDA" ? 0.083 : 0.034,
    relative_strength_spy: selected.symbol === "NVDA" ? 0.034 : 0.006,
    volume_ratio_20d: 1.32,
    signal: matchedSignal,
    price_history: Array.from({ length: 48 }, (_, index) => ({
      date: isoDate(47 - index),
      close: Number((120 + index * 0.75 + Math.sin(index / 3) * 4).toFixed(2)),
      volume: 15000000 + index * 50000,
    })),
    events: [
      {
        id: `demo-${selected.symbol}-earnings`,
        symbol: selected.symbol,
        signal_date: isoDate(1),
        event_type: "earnings",
        sentiment: 0.4,
        confidence: 0.55,
        source_url: "https://www.sec.gov/edgar/search/",
        summary: `${selected.symbol} demo earnings and filing watch item`,
      },
      {
        id: `demo-${selected.symbol}-news`,
        symbol: selected.symbol,
        signal_date: isoDate(2),
        event_type: "general",
        sentiment: 0.2,
        confidence: 0.45,
        source_url: "https://www.gdeltproject.org/data.html",
        summary: `${selected.symbol} demo market news item from free source discovery`,
      },
    ],
    what_changed: [
      "Demo momentum is constructive versus the watchlist baseline.",
      "The signal is awaiting real price, filing, and news ingestion.",
    ],
    why_it_matters: [
      "StockTek ranks tickers by probability, confidence, and evidence quality.",
      "Every real signal should be checked against source links and backtests.",
    ],
    risks: matchedSignal.risks,
    next_watch: [
      "Run Refresh Daily Data after market close.",
      "Check whether the signal still beats SPY in backtesting.",
    ],
    is_demo: true,
  };
}

export const demoBacktest: BacktestResponse = {
  id: "demo-backtest",
  created_at: new Date().toISOString(),
  universe: ["NVDA", "MSFT", "AMD"],
  strategy_rules: {
    threshold: 0.55,
    top_n: 5,
    horizon_days: 5,
    transaction_cost_bps: 5,
  },
  benchmark: "SPY",
  metrics: {
    trade_count: 0,
    win_rate: null,
    average_return: null,
    average_excess_return: null,
    max_drawdown: null,
    sharpe_like: null,
    calibration: "Run a real ingest first.",
  },
  trades: [],
};

function signal(symbol: string, probability: number, confidence: "low" | "medium" | "high"): Signal {
  return {
    id: `demo-${symbol}`,
    symbol,
    signal_date: isoDate(),
    horizon: "5 trading days",
    probability_outperform_spy: probability,
    confidence,
    score: Number(((probability - 0.5) * 2).toFixed(4)),
    drivers: [
      "Recent relative strength is positive in the demo dataset.",
      "Source sentiment skews slightly positive.",
    ],
    risks: [
      "This is demo data until a daily refresh completes.",
      "V1 signals are research aids, not trade instructions.",
    ],
    evidence: [
      {
        type: "filing",
        summary: `${symbol} demo SEC filing watch item`,
        url: "https://www.sec.gov/edgar/search/",
        sentiment: 0.2,
        confidence: 0.5,
      },
    ],
  };
}

